File: LeetCodePython/py3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        right = 0
        ans = 0
        dict = {}
        while right<len(s):
            pos = dict.get(s[right], -1)
            if pos<left:  # 未重复-1；或者重复值在滑动窗口左端边界外 
                dict[s[right]] = right
                ans = max(ans, right-left+1)
                right+=1
            else:
                left = pos+1  # 更新滑动窗口左边界
        return ans

if __name__ == '__main__':
    import time
    start = time.process_time()
    source = "abcabcbb"
    ans = Solution().lengthOfLongestSubstring(source)
    print(ans)
    end = time.process_time()
    print(str(end-start))

File: LeetCodeCPlusPlus/py3.cpp
#include<iostream>
#include<string>
#include<set>
using namespace std;
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        set<char> myset;
        int i=0,j=0, n = s.size(), ans=0, result=0;
        while(i<n and j<n) {
            if (myset.count(s[j])==0) {
                myset.insert(s[j++]);
                result = max(result,j-i);
            } else {
                myset.erase(s[i++]);
            }
        }
       return result;
    }
};

int main(){
    Solution solution = Solution();
    string source = "abcabcccd";
    cout << solution.lengthOfLongestSubstring(source)<<endl;
    return 0;
}